File: llrbtree/typescript/llrbtree.ts
export type Comparator<T> = (t1: T, t2: T) => -1 | 0 | 1;

const DEFAULT_COMPARATOR: Comparator<any> = (a, b) => {
  return a < b ? -1 : a === b ? 0 : 1;
}

export interface LLRBNode<T> {
  item: T;
  left: LLRBNode<T> | null;
  right: LLRBNode<T> | null;
  black: boolean;
}

function isRed<T>(h: LLRBNode<T> | null): boolean {
  if (h == null) {
    return false;
  }

  return !h.black;
}

export class LLRBTree<T> {
  count = 0;
  root: LLRBNode<T> | null = null;

  constructor(
    private readonly comparator: Comparator<T> = DEFAULT_COMPARATOR
  ) {}

  getRoot(): LLRBNode<T> | null {
    return this.root;
  }

  getLen(): number {
    return this.count;
  }

  // Get retrieves an element from the tree whose order is the same as that of key.
  get(key: T): T | null {
    let h = this.root;
    while (h != null) {
      const compare = this.comparator(key, h.item);
      switch (compare) {
        // case key < h.item:
        case -1: {
          h = h.left;
          break;
        }
        // case key > h.item:
        case 1: {
          h = h.right;
          break;
        }
        // key = h.item
        default:
          return h.item;
      }
    }

	  return null;
  }

  insert(item: T | null) {
    if (item == null) {
      throw new Error('inserting null item');
    }
    this.root = this.insertNode(this.root, item);
    this.root.black = true;
    this.count++;
  }

  private insertNode(h: LLRBNode<T> | null, item: T): LLRBNode<T> {
    if (h == null) {
      return { item, left: null, right: null, black: false };
    }
  
    // if item.Less(h.Item) {
    if (this.comparator(item, h.item) < 0) {
      // item < h.item
      h.left = this.insertNode(h.left, item);
    } else {
      h.right = this.insertNode(h.right, item);
    }
  
    if (isRed(h.right) && !isRed(h.left)) {
      h = this.rotateLeft(h);
    }
  
    if (isRed(h.left) && isRed(h.left!.left)) {
      h = this.rotateRight(h);
    }
  
    if (isRed(h.left) && isRed(h.right)) {
      this.flip(h);
    }
  
    return h;
  }

  // DeleteMin deletes the minimum element in the tree and returns the
  // deleted item or null otherwise.
  deleteMin(): T | null {
    let deleted: T | null;
    [this.root, deleted] = this.deleteMinNode(this.root);
    if (this.root != null) {
      this.root.black = true;
    }
    if (deleted != null) {
      this.count--;
    }

    return deleted;
  }

  // deleteMin code for LLRBTree 2-3 trees
  private deleteMinNode(h: LLRBNode<T> | null): [LLRBNode<T> | null, T | null] {
    if (h == null) {
      return [null, null];
    }
    if (h.left == null) {
      return [null, h.item];
    }

    if (!isRed(h.left) && !isRed(h.left.left)) {
      h = this.moveRedLeft(h);
    }

    let deleted: T | null;
    [h.left, deleted] = this.deleteMinNode(h.left);

    return [this.fixUp(h), deleted];
  }

  // DeleteMax deletes the maximum element in the tree and returns
  // the deleted item or null otherwise
  deleteMax(): T | null {
    let deleted: T | null;
    [this.root, deleted] = this.deleteMaxNode(this.root);
    if (this.root != null) {
      this.root.black = true;
    }
    if (deleted != null) {
      this.count--;
    }

    return deleted;
  }

  private deleteMaxNode(h: LLRBNode<T> | null): [LLRBNode<T> | null, T | null] {
    if (h == null) {
      return [null, null];
    }
    if (isRed(h.left)) {
      h = this.rotateRight(h);
    }
    if (h.right == null) {
      return [null, h.item]
    }
    if (!isRed(h.right) && !isRed(h.right.left)) {
      h = this.moveRedRight(h);
    }
    let deleted: T | null;
    [h.right, deleted] = this.deleteMaxNode(h.right);

    return [this.fixUp(h), deleted];
  }

  // // Delete deletes an item from the tree whose key equals key.
  // // The deleted item is return, otherwise null is returned.
  Delete(key: T): T | null {
    let deleted: T | null;
    [this.root, deleted] = this.deleteNode(this.root, key);
    if (this.root != null) {
      this.root.black = true
    }
    if (deleted != null) {
      this.count--;
    }

    return deleted;
  }

  private deleteNode(h: LLRBNode<T> | null, item: T): [LLRBNode<T> | null, T | null] {
    let deleted: T | null;
    if (h == null) {
      return [null, null];
    }
    if (this.comparator(item, h.item) < 0) {
      // item < h.item
      if (h.left == null) { // item not present. Nothing to delete
        return [h, null];
      }
      if (!isRed(h.left) && !isRed(h.left.left)) {
        h = this.moveRedLeft(h);
      }
      [h.left, deleted] = this.deleteNode(h.left, item);
    } else {
      if (isRed(h.left)) {
        h = this.rotateRight(h);
      }
      // If @item equals @h.Item and no right children at @h
      if (this.comparator(item, h.item) === 0 && h.right == null) {
        return [null, h.item];
      }
      // PETAR: Added 'h.right != null' below
      if (h.right != null && !isRed(h.right) && !isRed(h.right.left)) {
        h = this.moveRedRight(h);
      }
      // If @item equals @h.Item, and (from above) 'h.right != null'
      if (this.comparator(item, h.item) === 0) {
        let subDeleted: T | null;
        [h.right, subDeleted] = this.deleteMinNode(h.right);
        if (subDeleted == null) {
          throw new Error('logic');
        }
        [deleted, h.item] = [h.item, subDeleted];
      } else { // Else, @item is bigger than @h.Item
        [h.right, deleted] = this.deleteNode(h.right, item);
      }
    }

    return [this.fixUp(h), deleted];
  }
  
  private rotateLeft(h: LLRBNode<T>): LLRBNode<T> {
    const x = h.right;
    if (x == null || x.black) {
      throw new Error('rotating a black link');
    }
    h.right = x.left;
    x.left = h;
    x.black = h.black;
    h.black = false;

    return x;
  }
  
  private rotateRight(h: LLRBNode<T>): LLRBNode<T> {
    const x = h.left;
    if (x == null || x.black) {
      throw new Error('rotating a black link');
    }
    h.left = x.right;
    x.right = h;
    x.black = h.black;
    h.black = false;

    return x;
  }
  
  // REQUIRE: left and right children must be present
  private flip(h: LLRBNode<T>) {
    if (h.left == null || h.right == null) {
      return;
    }
    h.black = !h.black;
    h.left.black = !h.left.black
    h.right.black = !h.right.black
  }
  
  // REQUIRE: left and right children must be present
  private moveRedLeft(h: LLRBNode<T>): LLRBNode<T> {
    if (h.left == null || h.right == null) {
      return h;
    }
    this.flip(h);
    if (isRed(h.right.left)) {
      h.right = this.rotateRight(h.right);
      h = this.rotateLeft(h);
      this.flip(h);
    }

    return h;
  }
  
  // REQUIRE: left and right children must be present
  private moveRedRight(h: LLRBNode<T>): LLRBNode<T> {
    if (h.left == null || h.right == null) {
      return h;
    }
    this.flip(h);
    if (isRed(h.left.left)) {
      h = this.rotateRight(h);
      this.flip(h);
    }

    return h;
  }

  private fixUp(h: LLRBNode<T>): LLRBNode<T> {
    if (isRed(h.right)) {
      h = this.rotateLeft(h);
    }
  
    if (isRed(h.left) && isRed(h.left!.left)) {
      h = this.rotateRight(h);
    }
  
    if (isRed(h.left) && isRed(h.right)) {
      this.flip(h);
    }
  
    return h;
  }
}

File: src/go/llrbtree/llrbtree.go
package llrbtree

// Item interface
type Item interface {
	Less(than Item) bool
}

// Node struct
type Node struct {
	Item
	Left, Right *Node
	Black       bool
}

// LLRBTree struct
type LLRBTree struct {
	count int
	root  *Node
}

// New allocates a new tree
func New() *LLRBTree {
	return &LLRBTree{}
}

// Root returns the root node of the tree.
// It is intended to be used by functions that serialize the tree.
func (t *LLRBTree) Root() *Node {
	return t.root
}

// Len returns the number of nodes in the tree.
func (t *LLRBTree) Len() int { return t.count }

// Has returns true if the tree contains an element whose order is the same as that of key.
func (t *LLRBTree) Has(key Item) bool {
	return t.Get(key) != nil
}

// Get retrieves an element from the tree whose order is the same as that of key.
func (t *LLRBTree) Get(key Item) Item {
	h := t.root
	for h != nil {
		switch {
		case key.Less(h.Item):
			h = h.Left
		case h.Item.Less(key):
			h = h.Right
		default:
			return h.Item
		}
	}
	return nil
}

// Insert inserts item into the tree. If an existing element
// has the same order, both elements remain in the tree.
func (t *LLRBTree) Insert(item Item) {
	if item == nil {
		panic("inserting nil item")
	}
	t.root = t.insert(t.root, item)
	t.root.Black = true
	t.count++
}

func (t *LLRBTree) insert(node *Node, item Item) *Node {
	if node == nil {
		return &Node{Item: item}
	}

	if item.Less(node.Item) {
		node.Left = t.insert(node.Left, item)
	} else {
		node.Right = t.insert(node.Right, item)
	}

	if isRed(node.Right) && !isRed(node.Left) {
		node = rotateLeft(node)
	}

	if isRed(node.Left) && isRed(node.Left.Left) {
		node = rotateRight(node)
	}

	if isRed(node.Left) && isRed(node.Right) {
		flip(node)
	}

	return node
}

// DeleteMin deletes the minimum element in the tree and returns the
// deleted item or nil otherwise.
func (t *LLRBTree) DeleteMin() Item {
	var deleted Item
	t.root, deleted = deleteMin(t.root)
	if t.root != nil {
		t.root.Black = true
	}
	if deleted != nil {
		t.count--
	}
	return deleted
}

// deleteMin code for LLRBTree 2-3 trees
func deleteMin(h *Node) (*Node, Item) {
	if h == nil {
		return nil, nil
	}
	if h.Left == nil {
		return nil, h.Item
	}

	if !isRed(h.Left) && !isRed(h.Left.Left) {
		h = moveRedLeft(h)
	}

	var deleted Item
	h.Left, deleted = deleteMin(h.Left)

	return fixUp(h), deleted
}

// DeleteMax deletes the maximum element in the tree and returns
// the deleted item or nil otherwise
func (t *LLRBTree) DeleteMax() Item {
	var deleted Item
	t.root, deleted = deleteMax(t.root)
	if t.root != nil {
		t.root.Black = true
	}
	if deleted != nil {
		t.count--
	}
	return deleted
}

func deleteMax(h *Node) (*Node, Item) {
	if h == nil {
		return nil, nil
	}
	if isRed(h.Left) {
		h = rotateRight(h)
	}
	if h.Right == nil {
		return nil, h.Item
	}
	if !isRed(h.Right) && !isRed(h.Right.Left) {
		h = moveRedRight(h)
	}
	var deleted Item
	h.Right, deleted = deleteMax(h.Right)

	return fixUp(h), deleted
}

// Delete deletes an item from the tree whose key equals key.
// The deleted item is return, otherwise nil is returned.
func (t *LLRBTree) Delete(key Item) Item {
	var deleted Item
	t.root, deleted = t.delete(t.root, key)
	if t.root != nil {
		t.root.Black = true
	}
	if deleted != nil {
		t.count--
	}
	return deleted
}

func (t *LLRBTree) delete(h *Node, item Item) (*Node, Item) {
	var deleted Item
	if h == nil {
		return nil, nil
	}
	if item.Less(h.Item) {
		if h.Left == nil { // item not present. Nothing to delete
			return h, nil
		}
		if !isRed(h.Left) && !isRed(h.Left.Left) {
			h = moveRedLeft(h)
		}
		h.Left, deleted = t.delete(h.Left, item)
	} else {
		if isRed(h.Left) {
			h = rotateRight(h)
		}
		// If @item equals @h.Item and no right children at @h
		if !h.Item.Less(item) && h.Right == nil {
			return nil, h.Item
		}
		// PETAR: Added 'h.Right != nil' below
		if h.Right != nil && !isRed(h.Right) && !isRed(h.Right.Left) {
			h = moveRedRight(h)
		}
		// If @item equals @h.Item, and (from above) 'h.Right != nil'
		if !h.Item.Less(item) {
			var subDeleted Item
			h.Right, subDeleted = deleteMin(h.Right)
			if subDeleted == nil {
				panic("logic")
			}
			deleted, h.Item = h.Item, subDeleted
		} else { // Else, @item is bigger than @h.Item
			h.Right, deleted = t.delete(h.Right, item)
		}
	}

	return fixUp(h), deleted
}

/* Internal Utilities Functions */

func isRed(node *Node) bool {
	if node == nil {
		return false
	}
	return !node.Black
}

func rotateLeft(h *Node) *Node {
	x := h.Right
	if x.Black {
		panic("rotating a black link")
	}
	h.Right = x.Left
	x.Left = h
	x.Black = h.Black
	h.Black = false
	return x
}

func rotateRight(h *Node) *Node {
	x := h.Left
	if x.Black {
		panic("rotating a black link")
	}
	h.Left = x.Right
	x.Right = h
	x.Black = h.Black
	h.Black = false
	return x
}

// REQUIRE: Left and Right children must be present
func flip(h *Node) {
	h.Black = !h.Black
	h.Left.Black = !h.Left.Black
	h.Right.Black = !h.Right.Black
}

// REQUIRE: Left and Right children must be present
func moveRedLeft(h *Node) *Node {
	flip(h)
	if isRed(h.Right.Left) {
		h.Right = rotateRight(h.Right)
		h = rotateLeft(h)
		flip(h)
	}
	return h
}

// REQUIRE: Left and Right children must be present
func moveRedRight(h *Node) *Node {
	flip(h)
	if isRed(h.Left.Left) {
		h = rotateRight(h)
		flip(h)
	}
	return h
}

func fixUp(h *Node) *Node {
	if isRed(h.Right) {
		h = rotateLeft(h)
	}

	if isRed(h.Left) && isRed(h.Left.Left) {
		h = rotateRight(h)
	}

	if isRed(h.Left) && isRed(h.Right) {
		flip(h)
	}

	return h
}

/* Interface int, string */

// Int Item interface
type Int int

// Less implements Item interface for Int
func (x Int) Less(than Item) bool {
	return x < than.(Int)
}

// String Item interface
type String string

// Less implements Item interface for String
func (x String) Less(than Item) bool {
	return x < than.(String)
}
